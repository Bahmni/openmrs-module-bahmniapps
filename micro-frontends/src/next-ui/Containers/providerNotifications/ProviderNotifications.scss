.provider-notifications-next-ui {
    padding: 10px;
    margin-top: 0;
    margin-bottom: 0;

    .provider-notification-headers {
      padding: 10px 0;
      display: flex;
      justify-content: space-between;
      align-items: center;
      font-size: 16px;
      color: black;
      font-weight: 500;
    }
  }

  