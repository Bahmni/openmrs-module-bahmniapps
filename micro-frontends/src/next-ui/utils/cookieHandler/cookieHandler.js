export const getCookies = () => {
    const cookies = document.cookie.split(';');
    const cookiesObj = {};
    cookies.forEach(cookie => {
        const cookieArr = cookie.split('=');
        cookiesObj[cookieArr[0].trim()] = decodeURIComponent(cookieArr[1]);
    });
    return cookiesObj;
}
