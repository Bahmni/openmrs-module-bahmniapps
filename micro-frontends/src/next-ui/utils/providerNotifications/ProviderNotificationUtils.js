import axios from "axios";
import { EMERGENCY_MEDICATIONS_BASE_URL, GET_DRUG_ACKNOWLEDGEMENT_URL, GET_PROVIDER_UUID_URL } from "../../constants";
import { parseDateArray } from "../utils";

export const getPatientIPDDashboardUrl = (patientUuid, visitUuid) =>
    `/bahmni/clinical/index.html#/default/patient/${patientUuid}/dashboard/visit/ipd/${visitUuid}?source=clinical`;

export const getEmergencyDrugAcknowledgements = async (locationUuid, property, providerUuid) => {
    const apiURL = GET_DRUG_ACKNOWLEDGEMENT_URL.replace("{location_uuid}", locationUuid).replace("{property}", property).replace("{provider_uuid}",providerUuid);

    try {
        const response = await axios.get(apiURL);
        if (response.status === 200) {
            return response.data;
        }
        return [];
    } catch (error) {
        console.error(error);
    }
}

export const sortMedicationList = (medicationList) => {
    medicationList.forEach(group => {
        group.sort((a, b) => parseDateArray(a.administered_date_time).diff(parseDateArray(b.administered_date_time)));
    });
    medicationList.sort((a, b) => parseDateArray(a[0].administered_date_time).diff(parseDateArray(b[0].administered_date_time)));
    return medicationList;
};

export const acknowledgeEmergencyMedication = async (emergencyMedication, medicationAdministrationUuid) => {
    try {
        const response = await axios.put(
            EMERGENCY_MEDICATIONS_BASE_URL.replace("{medication_administration_uuid}",medicationAdministrationUuid),
            emergencyMedication
        );
        if (response.status === 200) {
            return { success: true, message: 'Emergency Medication Acknowledged Successfully' };
        } else {
            const errorMessage = response.data.error.message;
            return { success: false, message: `Error: ${errorMessage}` };
        }
    } catch (error) {
        return { success: false, message: `Error: ${error}` };
    }
};

export const getProvider = async () => {
    try {
        const response = await axios.get(GET_PROVIDER_UUID_URL);
        if (response.status === 200) {
            return response.data;
        }
        return [];
    } catch (error) {
        console.error(error);
    }
}

export const groupByIdentifier = (data, key) => {
    return data.reduce((accumulator, item) => {
      if (!accumulator[item[key]]) {
        accumulator[item[key]] = [];
      }
      accumulator[item[key]].push(item);
      return accumulator;
    }, {});
  };
  