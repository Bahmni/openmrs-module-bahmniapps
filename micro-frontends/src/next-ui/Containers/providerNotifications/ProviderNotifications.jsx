import React, { useEffect, useState } from 'react';
import './ProviderNotifications.scss';
import PropTypes from 'prop-types';
import PatientsList from '../../Components/ProviderNotificationPatients/PatientsList';
import { FormattedMessage } from 'react-intl';
import { I18nProvider } from '../../Components/i18n/I18nProvider';
import { getCookies } from '../../utils/cookieHandler/cookieHandler';
import { getProvider, getEmergencyDrugAcknowledgements, sortMedicationList, groupByIdentifier } from '../../utils/providerNotifications/ProviderNotificationUtils';
import { acknowledgeEmergencyMedication } from '../../utils/providerNotifications/ProviderNotificationUtils';
import { MEDICATION_ACKNOWLEDGE_SQL_PROPERTY, verifierFunction } from '../../constants';
import { NotificationCarbon } from "bahmni-carbon-ui";

export function ProviderNotifications() {
    const acknowledgementRequiredText = (<FormattedMessage id="AKNOWLEDGEMENT_REQUIRED_TEXT" defaultMessage="Acknowledgement required" />);
    const noDrugsToBeAcknowledgedText = (<FormattedMessage id="NO_DRUGS_TO_BE_AKNOWLEDGED_TEXT" defaultMessage="You have no new drugs to be acknowledged." />);
    const [patientListWithMedications, setPatientListWithMedications] = useState([]);
    const [showNotification, setShowNotification] = useState(false);
    const [notificationTitle, setNotificationTitle] = useState("");
    const [notificationType, setNotificationType] = useState("success");
    const [providerUuid, setProviderUuid] = useState();
    const cookies = getCookies();
    const locationUuid = JSON.parse(cookies["bahmni.user.location"]).uuid;

    const fetchProviderAndMedications = async () => {
        try {
            const provider = await getProvider();
            const providerUuid = provider.currentProvider.uuid;
            setProviderUuid(providerUuid);
            const emergencyDrugAcknowledgementResponse = await getEmergencyDrugAcknowledgements(locationUuid, MEDICATION_ACKNOWLEDGE_SQL_PROPERTY, providerUuid);
            const patientListGroupedByIdentifier = groupByIdentifier(emergencyDrugAcknowledgementResponse, 'identifier');
            const sortedMedicationList = sortMedicationList(Object.values(patientListGroupedByIdentifier));
            setPatientListWithMedications(sortedMedicationList);
        } catch (error) {
            console.error(error);
        }
    };
    
    const handleOnClick = async (medication_administration_performer_uuid, medication_administration_uuid, providerNotes) => {
        try {
            const response = await acknowledgeEmergencyMedication({
                providers: [
                {
                    uuid: medication_administration_performer_uuid,
                    providerUuid,
                    function: verifierFunction
                }
                ],
                notes: [{ authorUuid: providerUuid, text: providerNotes }]
            }, medication_administration_uuid);
            updateNotification(response)
            fetchProviderAndMedications();
        } catch (error) {
            updateNotification({ success: false, message: error.message });
        }
    };

    const updateNotification = async (response) => {
        setShowNotification(true);
        setNotificationTitle(response.message);
        setNotificationType(response.success ? "success" : "error");
    }

    const hideNotification = () => {
        setShowNotification(false);
    }

    useEffect(() => {
        fetchProviderAndMedications();
    }, []);

    const notificationComponent = (
        <NotificationCarbon 
            messageDuration={3000} 
            onClose={hideNotification} 
            showMessage={showNotification} 
            kind={notificationType} 
            title={notificationTitle}
        />
    );

    return (
        <I18nProvider>
            <div className='provider-notifications-next-ui'>
                {patientListWithMedications.length === 0 ? (
                    <div className='provider-notification-headers'>{noDrugsToBeAcknowledgedText}</div>
                ) : (
                    <>
                        <div className='provider-notification-headers'>
                            <span>{acknowledgementRequiredText}: </span>
                        </div>
                        <PatientsList patientListWithMedications={patientListWithMedications} handleOnClick={handleOnClick} />
                    </>
                )}
                {notificationComponent}
            </div>
        </I18nProvider>
    );
}

ProviderNotifications.propTypes = {
    hostData: PropTypes.object,
    hostApi: PropTypes.object
};
