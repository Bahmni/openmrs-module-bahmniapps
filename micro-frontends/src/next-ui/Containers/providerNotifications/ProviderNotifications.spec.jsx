import React from 'react';
import { render, waitFor, screen, fireEvent, act } from '@testing-library/react';
import { ProviderNotifications } from './ProviderNotifications';
import { getCookies } from '../../utils/cookieHandler/cookieHandler';
import { getProvider, getEmergencyDrugAcknowledgements, sortMedicationList, groupByIdentifier, acknowledgeEmergencyMedication, getPatientIPDDashboardUrl } from '../../utils/providerNotifications/ProviderNotificationUtils';

jest.mock('../../utils/cookieHandler/cookieHandler', () => ({
    getCookies: jest.fn(),
}));

jest.mock('../../utils/providerNotifications/ProviderNotificationUtils', () => ({
    getProvider: jest.fn(),
    getEmergencyDrugAcknowledgements: jest.fn(),
    sortMedicationList: jest.fn(),
    groupByIdentifier: jest.fn(),
    acknowledgeEmergencyMedication: jest.fn(),
    getPatientIPDDashboardUrl: jest.fn()
}));

describe('ProviderNotifications Component', () => {
    const cookiesMock = {
        "bahmni.user.location": JSON.stringify({ uuid: 'location-uuid' }),
    };
    const providerMock = { currentProvider: { uuid: 'provider-uuid' } };
    const emergencyDrugAcknowledgementResponseMock = [
        {
        identifier: 'PID123',
        administered_date_time: [2022,6,9],
        administered_drug_name: 'Drug A',
        administered_dose: '10mg',
        administered_dose_units: 'mg',
        administered_route: 'Oral',
        date_of_birth: [1990,1,1],
        gender: 'M',
        name: 'John Doe',
        patient_uuid: 'patient123',
        visit_uuid: 'visit123',
        medication_administration_performer_uuid: 'performer123',
        medication_administration_uuid: 'medication123'
        }
    ];
    const groupedByIdentifierMock = { PID123: emergencyDrugAcknowledgementResponseMock };
    const sortedMedicationListMock = [emergencyDrugAcknowledgementResponseMock];
    const acknowledgeResponseMock = { success: true, message: 'Acknowledgement successful' };

    beforeEach(() => {
        jest.clearAllMocks();
        jest.useFakeTimers().setSystemTime(new Date('2024-06-09'));
        getCookies.mockReturnValue(cookiesMock);
        getProvider.mockResolvedValue(providerMock);
        getEmergencyDrugAcknowledgements.mockResolvedValue(emergencyDrugAcknowledgementResponseMock);
        groupByIdentifier.mockReturnValue(groupedByIdentifierMock);
        sortMedicationList.mockReturnValue(sortedMedicationListMock);
        acknowledgeEmergencyMedication.mockResolvedValue(acknowledgeResponseMock);
    });

    it('should render without crashing and display no drugs to be acknowledged message', async () => {
        jest.clearAllMocks();
        sortMedicationList.mockReturnValue([]);
        await act(async () => {
            render(<ProviderNotifications />);
        });
        expect(await screen.findByText(/You have no new drugs to be acknowledged/i)).toBeTruthy();
    });

    it('renders the component without crashing when drugs are to be acknowledged', async () => {
        let container;
        await act(async () => {
            ({ container } = render(<ProviderNotifications />));
        });
        expect(container).toMatchSnapshot();
    });

    it('should render and display acknowledgment required message with patients list', async () => {
        await act(async () => {
            render(<ProviderNotifications />);
        });
        expect(await screen.findByText(/Acknowledgement required:/i)).toBeTruthy();
        expect(await screen.findByText(/John Doe - Male, 34 years 4 months 8 days/i)).toBeTruthy();
    });

    it('should call handleOnClick and display success notification on acknowledge button click', async () => {
        await act(async () => {
            render(<ProviderNotifications />);
        });
        await waitFor(() => expect(screen.getByText(/John Doe - Male, 34 years 4 months 8 days/i)).toBeTruthy());
        const notesInput = screen.getByPlaceholderText('Enter Notes');
        act(() => {
            fireEvent.change(notesInput, { target: { value: 'Some notes' } });
        });
        expect(notesInput.value).toBe('Some notes');
        const acknowledgeButton = screen.getByText("Acknowledge", { exact: true });
        act(() => {
            fireEvent.click(acknowledgeButton);
        });
        await waitFor(() => expect(acknowledgeEmergencyMedication).toHaveBeenCalled());
        expect(await screen.findByText(/Acknowledgement successful/i)).toBeTruthy();
    });

    it('should display error notification if acknowledgeEmergencyMedication fails', async () => {
        const errorResponseMock = { success: false, message: 'Acknowledgement failed' };
        acknowledgeEmergencyMedication.mockResolvedValueOnce(errorResponseMock);
        await act(async () => {
            render(<ProviderNotifications />);
        });
        await waitFor(() => expect(screen.getByText(/John Doe - Male, 34 years 4 months 8 days/i)).toBeTruthy());
        const notesInput = screen.getByPlaceholderText('Enter Notes');
        act(() => {
            fireEvent.change(notesInput, { target: { value: 'Some notes' } });
        });
        expect(notesInput.value).toBe('Some notes');
        const acknowledgeButton = screen.getByText("Acknowledge", { exact: true });
        act(() => {
            fireEvent.click(acknowledgeButton);
        });
        await waitFor(() => expect(acknowledgeEmergencyMedication).toHaveBeenCalled());
        expect(await screen.findByText(/Acknowledgement failed/i)).toBeTruthy();
    });

    it('should fetch provider and medications on component mount', async () => {
        await act(async () => {
            render(<ProviderNotifications />);
        });
        await waitFor(() => expect(getProvider).toHaveBeenCalled());
        await waitFor(() => expect(getEmergencyDrugAcknowledgements).toHaveBeenCalled());
    });

    it('should handle errors during fetching provider and medications', async () => {
        getProvider.mockRejectedValueOnce(new Error('Failed to fetch provider'));
        console.error = jest.fn(); // Suppress error logging in test output
        await act(async () => {
            render(<ProviderNotifications />);
        });
        await waitFor(() => expect(getProvider).toHaveBeenCalled());
        expect(console.error).toHaveBeenCalledWith(new Error('Failed to fetch provider'));
    });

    it('should set notification when an error occurs', async () => {
        acknowledgeEmergencyMedication.mockImplementationOnce(() => {
            throw new Error('New Error');
        });
        await act(async () => {
            render(<ProviderNotifications />);
        });
        await waitFor(() => expect(screen.getByText(/John Doe - Male, 34 years 4 months 8 days/i)).toBeTruthy());
        const notesInput = screen.getByPlaceholderText('Enter Notes');
        act(() => {
            fireEvent.change(notesInput, { target: { value: 'Some notes' } });
        });
        expect(notesInput.value).toBe('Some notes');
        const acknowledgeButton = screen.getByText("Acknowledge", { exact: true });
        act(() => {
            fireEvent.click(acknowledgeButton);
        });
        expect(screen.getByText(/New Error/i)).toBeTruthy();
        const notification = screen.getByRole('alert');
        expect(notification.getAttribute("class")).toContain("error");
    });

    it('should set showNotification to false when NotificationCarbon onClose is called', async () => {
        await act(async () => {
            render(<ProviderNotifications />);
        });
        await waitFor(() => expect(screen.getByText(/John Doe - Male, 34 years 4 months 8 days/i)).toBeTruthy());
        const notesInput = screen.getByPlaceholderText('Enter Notes');
        act(() => {
            fireEvent.change(notesInput, { target: { value: 'Some notes' } });
        });
        expect(notesInput.value).toBe('Some notes');
        const acknowledgeButton = screen.getByText("Acknowledge", { exact: true });
        jest.useRealTimers();
        act(() => {
            fireEvent.click(acknowledgeButton);
        });
        await waitFor(() => expect(screen.getByText(/Acknowledgement successful/i)).toBeTruthy());
        await new Promise(res => setTimeout(res, 3000));
        expect(screen.queryByText(/Acknowledgement successful/i)).not.toBeTruthy();
    });
});
