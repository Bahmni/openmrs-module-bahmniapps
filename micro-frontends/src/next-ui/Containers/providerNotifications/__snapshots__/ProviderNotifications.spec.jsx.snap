// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProviderNotifications Component renders the component without crashing when drugs are to be acknowledged 1`] = `
<div>
  <div
    class="provider-notifications-next-ui"
  >
    <div
      class="provider-notification-headers"
    >
      <span>
        Acknowledgement required
        : 
      </span>
    </div>
    <ul
      class="bx--accordion patient-list-dropdown bx--accordion--end bx--accordion--md"
    >
      <li
        class="bx--accordion__item"
      >
        <button
          aria-controls="accordion-item-2"
          aria-expanded="false"
          class="bx--accordion__heading"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="bx--accordion__arrow"
            fill="currentColor"
            focusable="false"
            height="16"
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 8L6 13 5.3 12.3 9.6 8 5.3 3.7 6 3z"
            />
          </svg>
          <div
            class="bx--accordion__title"
            dir="auto"
          >
            <div
              class="patient-list-tile-content"
            >
              <div
                class="warning"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  focusable="false"
                  height="16"
                  preserveAspectRatio="xMidYMid meet"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M14.5,14h-13c-0.2,0-0.3-0.1-0.4-0.2c-0.1-0.2-0.1-0.3,0-0.5l6.5-12C7.7,1,8,0.9,8.2,1.1c0.1,0,0.2,0.1,0.2,0.2l6.5,12	c0.1,0.2,0.1,0.3,0,0.5C14.9,13.9,14.7,14,14.5,14z M2.3,13h11.3L8,2.5L2.3,13z"
                  />
                  <path
                    d="M7.5 6H8.5V9.5H7.5zM8 10.8c-.4 0-.8.3-.8.8s.3.8.8.8c.4 0 .8-.3.8-.8S8.4 10.8 8 10.8z"
                  />
                </svg>
                <span
                  class="drug-count"
                >
                  1
                </span>
              </div>
              <div
                class="patient-info"
              >
                <a
                  class="bx--link patient-id"
                >
                  (PID123)
                </a>
                <span>
                  John Doe - Male, 34 years 4 months 8 days
                </span>
              </div>
            </div>
          </div>
        </button>
        <div
          class="bx--accordion__content"
          id="accordion-item-2"
        >
          <div
            class="patient-list-content"
          >
            <span>
              9/6/2022
            </span>
            <div
              class="content-info"
            >
              <span>
                Drug A
              </span>
              <div
                class="notes"
              >
                <div
                  class="bx--form-item"
                >
                  <div
                    class="bx--text-area__label-wrapper"
                  >
                    <label
                      class="bx--label"
                    >
                      <div
                        class="titleText"
                      >
                        <span>
                          Note 
                        </span>
                        <span
                          class="required"
                        >
                          *
                        </span>
                      </div>
                    </label>
                  </div>
                  <div
                    class="bx--text-area__wrapper"
                  >
                    <textarea
                      class="bx--text-area patient-list-text-area"
                      cols="50"
                      placeholder="Enter Notes"
                      required=""
                      rows="1"
                    />
                  </div>
                </div>
                <button
                  class="patient-list-button bx--btn bx--btn--primary bx--btn--disabled"
                  disabled=""
                  tabindex="0"
                  type="button"
                >
                  Acknowledge
                </button>
              </div>
            </div>
          </div>
        </div>
      </li>
    </ul>
    <div
      class="alertContainer"
    />
  </div>
</div>
`;
